README.rst
===============================
SoftClusterEX
===============================

    Toolname:    Soft Cluster EX
    Author:      Webber Huang
    Contact:     xracz.fx@gmail.com
    Homepage:    http://riggingtd.com
    Version:     3.0.0
    Requires:    Maya 2011 x64 or above
    Platform:    Windows 64bit, Mac OS X, Linux 64bit
    License:     BSD license


Description:
----------------------------------------------
    Soft Cluster EX is a python and c++ based tool for converting soft selection to
    cluster or joint, supports common types of deformable geometry(e.g. polygon,nurbs
	,subdiv and lattice) with a lot of advance features.


Install:
----------------------------------------------
    1. Place SoftClusterEX folder into your maya script path.
       e.g., C:/Users/YourUsername/Documents/maya/scripts/
       
       * NOTE 1: Please remove softClusterEX.py and softClusterEX.pyc first if you ever 
	             installed the 1.x version before.
	   * NOTE 2: Please remove the softSelectionQuery plugins from your plugin path 
	             (e.g., C:/Program Files/Autodesk/Maya<Version>/bin/plug-ins/) if you 
				 ever installed 2.6.0 version or older.
			   
    2. Drag "install.mel" into maya scene, a new button will appear in your current shelf, 
       click it to launch. (Recommend!)


Usage:
----------------------------------------------
If you don't like to use "install.mel" to help you install, run the following code in maya's
python tab is also okay. 

    import SoftClusterEX
    SoftClusterEX.launch()


Features:
----------------------------------------------
#### General functionality:

  - supports create cluster or joint from polygon, nurbsSurface, nurbsCurve, subdiv and lattice
  - all types can be toggled on/off from UI
  - supports create cluster or joint from multi objects of different type
  - change falloff mode
  - supports exclude influenced objects and groups while working on Global falloff mode
  - supports objects with duplicate name
  - ultra high performance, even working on massive density geometry
  - mirror cluster or joint in local or world space, also support working on
    multiple types of deformable geometry
  - redefine weight with soft selection
  - supports export and import weight

#### Command line:

  - The core function of this tool comes from the custom command "softSelectionQuery",
    you can use it to implement your own tools.

  - Synopsis: softSelectionQuery [flags] [String...]
      - sl -selection                            return a list contains transforms and vertices
      - tr -transforms                           return transform names of influenced objects
      - s -shapeNames:                           return shape names of influenced objects 
      - vtx -vertices:                           return vertices of influenced objects  
      - ap -allPaths:                            return partial paths to objects in DAG
      - l -long:                                 return full path names for Dag objects.
      - c -count:                                return count of influenced objects          
      - nt -nodeTypes:                           return type names of influenced objects
      - ant -apiNodeTypes:                       return api type name of influenced objects  
      - exo -excludeObjects  String (multi-use): set exclude objects	  
      - t -types           String (multi-use):   set support types
      - w -weights:                              return weights of influenced components


Todo:
----------------------------------------------
  - None

  
Limitation
----------------------------------------------
  - Joint function can't work with subdiv since MFnSkinCluster incompatible with this type of geometry


Feedback:
----------------------------------------------
Bugs, questions and suggestions to xracz.fx@gmail.com

HISTORY.rst
History
----------------------------------------------
#### 2014-09-07: v3.0.0 by Webber Huang
  - NEW: Fully supports working with joint, create, mirror,
    redefine, export, import weight, just like play with cluster
  - NEW: Convert cluster and joint into each other
  - NEW: Append new weight with redefine tool
  - NEW: A falloff curve panel to help user adjust soft selection
  - UPDATE: Better UI
  
#### 2014-09-03: v2.7.0 by Webber Huang
  - Reduce the redundant codes, improve mirror performance.
  - Reduce the information contained in weight data file.
  - A new method for setting environment, plugins have been 
    integrated into "SoftClusterEX" module, so people don't 
	need to tangle with placing the plugin files any more.
  - NEW: A mel script to help people install a shelf button.
  - UPDATE: softSelectionQuery plugin has been update to v1.5,
    new ways to deal with the path of the geometry vertex.
  - FIXED: Failure to launch in maya 2011.
  - FIXED: Malfunction occured in mirror operation.
  
#### 2014-08-28: v2.6.0 by Webber Huang
  - Reconstruct source files to make it more reasonable
  - Simplify weight data structure, performance improve
  - NEW: Redefine cluster weight by soft selection
  - NEW: Detach influenced geometries from cluster 
  - NEW: Export and Import cluster weight
  - NEW: A flat style icon 

#### 2014-08-17: v2.1.0 by Webber Huang
  - Optimized architecture to meet fully OO design principle
  - Optimized UI
  - A new tab to show author's information
  
#### 2014-07-25: v2.0.0 by Webber Huang
  - A complete redisign of architecture, optimized codes and algorithms
  
  - Base on Maya API and optimized data structure, this version achieves a significant
	performance improvement when create cluster on massive density geometry

  - All geometry vertices are binded to cluster, it means that no more necessary to use
	"Edit Membership Tool" to append vertices to cluster, if you want some more
	influence, just paint it
	
  - A powerful mirror function has been implemented to enhance this tool, supports
	all kind of common deformable geometry, base on kd-tree algorithm, it works fast
	and reliable, so please feel free to play it with your wide imagination
	
  - Notes that there may be a few vertices mis-binded(zero weight value) after perform
	mirror operation, it happens when geometry is not perfect symmetry, i can't solve
	this issue in current, please paint it when you come across this situation, i hope
	it won't be such a troublesome problem to you, this function works perfect mostly
  
#### 2014-05-28: v1.1.1 by Webber Huang
  - Nothing's news, but support maya 2015 SP2 and update compile against 2014 SP4

#### 2013-11-18: v1.1.0 by Webber Huang
  - Add Linux compile version for softSelectionQuery plugin, it works on linux now

  - Exclude objects by group supported, user can add groups into exclude list,
	objects under these groups will be ignored while working on global mode

#### 2013-10-27: v1.02 by Webber Huang
  - Compiled softSelectionQuery plugin for OS X, mac user can play with it

#### 2013-10-09: v1.01 by Webber Huang
  - Issue of plugin loading failed,i re-compile all softSelectionQuery_<Version>-<Bit>.mll
	with vs2008 to maintain maximum compatibility

#### 2013-09-30: v1.00 by Webber Huang
  - Initial release

SoftClusterEX/ui/tabCreate.py
"""
#----------------------------------------------------------------------
#    This file is part of "Soft Cluster EX"
#    and covered by a BSD-style license, check
#    LICENSE for detail.
#
#    Author:      Webber Huang
#    Contact:     xracz.fx@gmail.com
#    Homepage:    http://riggingtd.com
#----------------------------------------------------------------------
"""
TAB = ''
ENABLE = True
INDEX = 1

from time import clock
import maya.cmds as cmds
import maya.mel as mel
import maya.OpenMaya as om
from SoftClusterEX.core import scData
from SoftClusterEX.startup import setup
import baseTab
reload(scData)
reload(baseTab)
reload(setup)


########################################################################
class TabCreate(baseTab.BaseTab):
    """"""
    _TITLE = 'Create'
    _INDEX = 1

    #----------------------------------------------------------------------
    def __init__(self):
        super(TabCreate, self).__init__()
        self.supportTypeCB = []
        self.checkAllCB = ''
        self.excludeScrollList = ''
        self.falloffModeOption = ''
        self.gradient = ''
        self.interpOption = ''
        self.excludeItems = ''
        self._supportTypes = setup.getSupportTypes()
        self.keepCheck = ''
        self.jointCheck = ''

    #----------------------------------------------------------------------
    def createUI(self, parent):
        """"""
        layout = cmds.frameLayout( lv=0, bv=0, cll=False, mh=4, mw=4, p=parent)
        
        form = cmds.formLayout()
        optLayout = self.installOption()

        # Install exclude object list
        excludeList = self.installExcludeObjectList()
        
        # Install falloff gradient
        gradient = self.installFalloffGradient()
        
        # Install create button
        createForm = self.installCreateButton()
        convertForm = self.installConvertButton()
        
        # Arrage UI elements
        cmds.formLayout(form,
                       e=1,
                       af=[(optLayout, 'left', 0),
                           (optLayout, 'right', 0),
                           (optLayout, 'top', 0),
                           (excludeList, 'left', 0),
                           (excludeList, 'right', 0),
                           (excludeList, 'top', 0),
                           (gradient, 'left', 0),
                           (gradient, 'right', 0),
                           (gradient, 'top', 0),
                           (createForm, 'top', 0), 
                           (createForm, 'bottom', 0), 
                           (createForm, 'right', 0),
                           (createForm, 'left', 0),
                           (convertForm, 'bottom', 0), 
                           (convertForm, 'right', 0),
                           (convertForm, 'left', 0)],
                       
                       ac=[(excludeList, 'top', 2, optLayout),
                           (gradient, 'top', 2, excludeList), 
                           (createForm, 'top', 2, gradient), 
                           (createForm, 'bottom', 2, convertForm)])
        
        cmds.setParent( '..' )
        cmds.setParent( '..' )
        return layout
    
    #----------------------------------------------------------------------
    def installOption(self):
        """"""
        layout = cmds.frameLayout( l='Options:', bs='etchedIn', cl=True, cll=True)
        # Install support types check boxes
        self.installSupportTypeCheckbox()
        
        # Install convert option
        self.installConvertOption()
        
        cmds.setParent( '..' )
        return layout
        
    #----------------------------------------------------------------------
    def installConvertOption(self):
        """"""
        cmds.rowLayout(nc=2, adj=2)
        cmds.text(l=' Convert:', fn='boldLabelFont')
        self.keepCheck = cmds.iconTextCheckBox( st='textOnly', l='Keep original', v=True )
        cmds.setParent( '..' )
        cmds.separator( style='none' )
        
    #----------------------------------------------------------------------
    def installSupportTypeCheckbox(self):
        """"""
        layout = cmds.frameLayout( l='Support Types:', borderStyle='etchedOut', cl=False, cll=False)
        cmds.rowColumnLayout(numberOfRows=2)

        for typeName in self._supportTypes:
            self.supportTypeCB.append(cmds.checkBox(l='%s' % self._supportTypes[typeName]))
            cmds.checkBox(self.supportTypeCB[-1], e=1, cc=lambda *args: self.toggleAllCB())
        
        self.checkAllCB = cmds.checkBox( l='All', cc=lambda *args: self.toggleAllSupportTypesCB())
        cmds.setParent( '..' )
        cmds.setParent( '..' )
        self.setDefaultSupportTypes()
        return layout

    #----------------------------------------------------------------------
    def installExcludeObjectList(self):
        """"""
        layout = cmds.frameLayout( l='Exclude Objects:', borderStyle='etchedIn', cl=True, cll=True)
        self.excludeScrollList = cmds.textScrollList(allowMultiSelection=True)
        cmds.popupMenu()
        cmds.menuItem(l='Add', c=lambda *args: self.addExcludeObjectsCmd())
        cmds.menuItem(l='Delete', c=lambda *args: self.deleteExcludeObjectsCmd())
        cmds.menuItem(l='Clear', c=lambda *args: self.clearExcludeObjectsCmd())
        cmds.setParent( '..' )
        return layout
    
    #----------------------------------------------------------------------
    def installFalloffOption(self):
        """"""
        self.falloffModeOption = cmds.optionMenuGrp( l='Falloff Mode:',
                                                     cc=lambda *args: self.changeFalloffMode(),
                                                     cl2=('left', 'left'),
                                                     cw=(1, 70),
                                                     adj=2 )
        cmds.menuItem( l='Volume' )
        cmds.menuItem( l='Surface' )
        cmds.menuItem( l='Global' )
        self.setDefaultFalloffMode()
        return self.falloffModeOption
    
    #----------------------------------------------------------------------
    def installInterpolationOption(self):
        """"""
        self.interpOption = cmds.optionMenuGrp(l='Interpolation:',
                                               cc=lambda *args: self.changeInterpolationValue(),
                                               cl2=('left', 'left'),
                                               cw=(1, 70),
                                               adj=2)
        cmds.menuItem( l='None' )
        cmds.menuItem( l='Linear' )
        cmds.menuItem( l='Smooth' )
        cmds.menuItem( l='Spline' )
        self.setDefaultInterpolation()
        return self.interpOption
    
    #----------------------------------------------------------------------
    def softSelectCurveKeyChanged(self):
        """"""
        interp = cmds.gradientControlNoAttr(self.gradient, q=1, civ=1)
        cmds.optionMenuGrp(self.interpOption, e=1, sl=interp+1)
    
    #----------------------------------------------------------------------
    def changeSoftSelectValue(self):
        curveValue = cmds.gradientControlNoAttr(self.gradient, q=1, asString=1)
        cmds.softSelect(e=1, ssc=curveValue)
    
    #----------------------------------------------------------------------
    def changeInterpolationValue(self):
        interp = cmds.optionMenuGrp(self.interpOption, q=1, sl=1)
        cmds.gradientControlNoAttr(self.gradient, e=1, civ=interp-1)
        
    #----------------------------------------------------------------------
    def installFalloffGradient(self):
        """"""
        layout = cmds.frameLayout( l='Falloff:', borderStyle='etchedIn', cl=True, cll=True, mw=5, mh=5)
        
        # Install falloff mode scroll List
        cmds.rowLayout(nc=2, adj=1)
        falloffMode = self.installFalloffOption()
        resetBtn = cmds.button(l='Reset', c=lambda *args: self.resetFalloff(), w=40)
        cmds.setParent( '..' )
        
        # Install falloff gradient control
        self.gradient = cmds.gradientControlNoAttr( 'falloffCurve', h=90)
        self.copyFromMayaFalloffCurve()
        cmds.gradientControlNoAttr( 'falloffCurve',
                                    e=True,
                                    optionVar='falloffCurveOptionVar',
                                    changeCommand=lambda *args: self.changeSoftSelectValue(),
                                    currentKeyChanged=lambda *args: self.softSelectCurveKeyChanged() )
        
        # Install interpolation scroll List
        interpolation = self.installInterpolationOption()
        
        cmds.setParent( '..' )
        return layout
    
    #----------------------------------------------------------------------
    def installCreateButton(self):
        """"""
        form = cmds.formLayout()
        btn = cmds.iconTextButton(st='iconOnly',
                                  l='Create',
                                  i=setup.getBanner(),
                                  c=lambda *args: self.createSoftDeformerCmd() )
        
        self.jointCheck = cmds.iconTextCheckBox(st='textOnly', l='   J   ',  ann="Create joint")
        cmds.formLayout(form,
                        e=1,
                        af=[(btn, 'left', 0),
                            (btn, 'right', 0),
                            (btn, 'top', 0),
                            (btn, 'bottom', 0),
                            (self.jointCheck, 'right', 0),
                            (self.jointCheck, 'top', 0)],
                        )
        cmds.setParent( '..' )
        return form
          
    #----------------------------------------------------------------------
    def installConvertButton(self):
        """"""
        return cmds.button( l='Convert',
                             ann='Cluster <-->  Joint', 
                             bgc=(153.0/255, 217.0/255, 234.0/255),
                             c=lambda *args: self.convertDeformerCmd() )

    #----------------------------------------------------------------------
    def addExcludeObjectsCmd(self):
        """"""
        sel = cmds.ls(sl=1, ap=1)
        self.excludeItems = cmds.textScrollList(self.excludeScrollList, q=1, ai=1)

        for obj in sel:
            if not self.excludeItems:
                cmds.textScrollList(self.excludeScrollList, e=1, a=obj)
            elif not obj in self.excludeItems:
                cmds.textScrollList(self.excludeScrollList, e=1, a=obj)

    #----------------------------------------------------------------------
    def clearExcludeObjectsCmd(self):
        """"""
        cmds.textScrollList(self.excludeScrollList, e=1, ra=1)

    #----------------------------------------------------------------------
    def deleteExcludeObjectsCmd(self):
        """"""
        selItems = cmds.textScrollList(self.excludeScrollList, q=1, si=1)
        cmds.textScrollList(self.excludeScrollList, e=1, ri=selItems)

    #----------------------------------------------------------------------
    def createSoftDeformerCmd(self):
        """"""
        self.excludeItems = cmds.textScrollList(self.excludeScrollList, q=1, ai=1)

        supportTypes = []
        for typeCB, typeName in zip(self.supportTypeCB, self._supportTypes):
            if cmds.checkBox(typeCB, q=1, v=1):
                supportTypes.append(typeName)

        if not supportTypes:
            cmds.error('Please check at least one support type!')
        
        start = clock()
        
        if not cmds.iconTextCheckBox(self.jointCheck, q=1, v=1):
            scData.ClusterFn.createSoftCluster(self.excludeItems, supportTypes)
        else:
            scData.JointFn.createSoftJoint(self.excludeItems, supportTypes)
            
        finish = clock()
        om.MGlobal.displayInfo( "Create Time Cost: %fs" % (finish-start) )
    
    #----------------------------------------------------------------------
    def convertDeformerCmd(self):
        """"""
        sels = cmds.ls(sl=1, ap=1)
        if not sels: cmds.error("Please select clusters or joints!")
            
        start = clock()
        for sel in sels:
            if cmds.objectType( sel, isType='joint' ):
                self.jointToCluster(sel)
            elif cmds.listRelatives(sel, s=1, typ='clusterHandle'):
                self.clusterToJoint(sel)
            else: cmds.warning("%s is skipped!" % sel)
            
            if not cmds.iconTextCheckBox(self.keepCheck, q=1, v=1):
                cmds.delete(sel)
        finish = clock()
        om.MGlobal.displayInfo( "Convert Time Cost: %fs" % (finish-start) )
            
    #----------------------------------------------------------------------
    def clusterToJoint(self, clusterHandle):
        """
        Args:
          clusterNode (str)
        """
        weightData = scData.ClusterFn.getWeightData(clusterHandle)
        pos = cmds.xform(clusterHandle, q=1, ws=1, piv=1)[0:3]
        scData.JointFn.createJoint(clusterHandle, pos, weightData)
        
    #----------------------------------------------------------------------
    def jointToCluster(self, joint):
        """
        Args:
          joint (str)
        """
        weightData = scData.JointFn.getWeightData(joint)
        pos = cmds.xform(joint, q=1, ws=1, t=1)
        scData.ClusterFn.createCluster(joint, pos, weightData, True)
            
    #----------------------------------------------------------------------
    def changeFalloffMode(self):
        """"""
        selIndex = cmds.optionMenuGrp(self.falloffModeOption, q=1, sl=1) - 1
        cmds.softSelect(ssf=selIndex)

    #----------------------------------------------------------------------
    def setDefaultFalloffMode(self):
        """"""
        cmds.optionMenuGrp(self.falloffModeOption, e=1, sl=2)
        cmds.softSelect(ssf=1)
    
    #----------------------------------------------------------------------
    def setDefaultInterpolation(self):
        cmds.optionMenuGrp(self.interpOption, e=1, sl=3)
    
    #----------------------------------------------------------------------
    def copyFromMayaFalloffCurve(self):
        """"""
        # Read maya's native soft select falloff curve
        curveValue = cmds.softSelect(q=1, ssc=1).split(',')
        curveValue = zip(curveValue[0::3], curveValue[1::3], curveValue[2::3])
        curveValue = [','.join(v) for v in curveValue]
        
        if cmds.optionVar( exists='falloffCurveOptionVar' ):
            cmds.optionVar( clearArray='falloffCurveOptionVar' )

        [cmds.optionVar( stringValueAppend=['falloffCurveOptionVar', v] ) for v in curveValue]
        
        cmds.gradientControlNoAttr( self.gradient, e=True, optionVar='falloffCurveOptionVar' )
    
    #----------------------------------------------------------------------
    def setDefaultFalloffCurve(self):
        """"""
        cmds.optionVar( clearArray='falloffCurveOptionVar' )
        cmds.optionVar(stringValueAppend=[ 'falloffCurveOptionVar', '0,1,2'])
        cmds.optionVar(stringValueAppend=[ 'falloffCurveOptionVar', '1,0,2'])
        cmds.gradientControlNoAttr( self.gradient, e=True, optionVar='falloffCurveOptionVar' )
        cmds.gradientControlNoAttr( self.gradient, e=True, asString='0,1,2,1,0,2' )
        cmds.softSelect(e=1, ssc='0,1,2,1,0,2')
    
    #----------------------------------------------------------------------
    def resetFalloff(self):
        """"""
        self.setDefaultFalloffMode()
        self.setDefaultInterpolation()
        self.setDefaultFalloffCurve()
        
    #----------------------------------------------------------------------
    def setDefaultSupportTypes(self):
        """"""
        for typeCB in self.supportTypeCB:
            cmds.checkBox(typeCB, e=1, v=1)

        cmds.checkBox(self.checkAllCB, e=1, v=1)

    #----------------------------------------------------------------------
    def toggleAllSupportTypesCB(self):
        """"""
        checkAll = cmds.checkBox(self.checkAllCB, q=1, v=1)
        for typeCB in self.supportTypeCB:
            cmds.checkBox(typeCB, e=1, v=checkAll)

    #----------------------------------------------------------------------
    def toggleAllCB(self):
        """"""
        enableAll = True
        for typeCB in self.supportTypeCB:
            if not cmds.checkBox(typeCB, q=1, v=1):
                enableAll = False
                break

        cmds.checkBox(self.checkAllCB, e=1, v=enableAll)


#----------------------------------------------------------------------
def getTab():
    """"""
    return TabCreate()

SoftClusterEX/ui/tabWeight.py
"""
#----------------------------------------------------------------------
#    This file is part of "Soft Cluster EX"
#    and covered by a BSD-style license, check
#    LICENSE for detail.
#
#    Author:      Webber Huang
#    Contact:     xracz.fx@gmail.com
#    Homepage:    http://riggingtd.com
#----------------------------------------------------------------------
"""

TAB = ''
ENABLE = True
INDEX = 3

import maya.cmds as cmds
import maya.OpenMaya as om
from SoftClusterEX.core import scData
from SoftClusterEX.startup import setup
import baseTab
reload(scData)
reload(baseTab)


########################################################################
class TabWeight(baseTab.BaseTab):
    """"""
    _TITLE = 'Weight'
    _INDEX = 3

    #----------------------------------------------------------------------
    def __init__(self):
        super(TabWeight, self).__init__()
        self.editBtn = ''
        self.doneBtn = ''
        self.appendCheck = ''
        self.detachBtn = ''
        self.influence = ''
        self._type = ''
        self.toggle = True
    
    #----------------------------------------------------------------------
    def createUI(self, parent):
        """"""
        layout = cmds.frameLayout( lv=0, bv=0, cll=False, mh=6, mw=4, p=parent)
        
        self.installRedefineSection()
        self.installExportImport()
        
        cmds.setParent( ".." )
        return layout
    
    #----------------------------------------------------------------------
    def installRedefineSection(self):
        """"""
        cmds.frameLayout( l='Redefine:', bs='etchedIn', cll=False, mh=5, w=200)
        
        form1 = cmds.formLayout()
        self.installEdit()
        self.installDone()
        self.installAppendCheck()
        
        cmds.formLayout(form1,
                       e=1,
                       af=[(self.editBtn, 'left', 0),
                           (self.editBtn, 'top', 0),
                           (self.editBtn, 'bottom', 0),
                           (self.doneBtn, 'left', 0),
                           (self.doneBtn, 'top', 0),
                           (self.doneBtn, 'bottom', 0),
                           (self.appendCheck, 'right', 0),
                           (self.appendCheck, 'top', 0),
                           (self.appendCheck, 'bottom', 0)],
                       ac=[(self.doneBtn, 'left', 1, self.editBtn),
                           (self.doneBtn, 'right', 4, self.appendCheck)], 
                       ap=[(self.editBtn, 'right', 0, 42)])
        
        cmds.setParent( ".." )
        
        separator = cmds.separator( st='in' )
        
        form2 = cmds.formLayout()
        self.installDetach()
        cmds.formLayout(form2,
                       e=1,
                       af=[(self.detachBtn, 'left', 0),
                           (self.detachBtn, 'top', 0),
                           (self.detachBtn, 'bottom', 0),
                           (self.detachBtn, 'right', 0)])
        cmds.setParent( ".." )
        cmds.setParent( ".." )
    
    #----------------------------------------------------------------------
    def installEdit(self):
        """"""
        self.editBtn =  cmds.button(l="Edit",
                                   bgc=(239.0/255, 238.0/255, 179.0/255), 
                                   en=self.toggle,
                                   c=lambda *args: self.editCmd() )
        
    
    #----------------------------------------------------------------------
    def installDone(self):
        """"""
        self.doneBtn = cmds.button(l="Done",
                                   ann="Select nothing to cancel redefine weight.", 
                                   bgc=(153.0/255, 217.0/255, 234.0/255), 
                                   en=not self.toggle,
                                   c=lambda *args: self.doneCmd() )
    
    #----------------------------------------------------------------------
    def installAppendCheck(self):
        """"""
        self.appendCheck = cmds.checkBox(l='A', ann='Append weights')
        
    #----------------------------------------------------------------------
    def installDetach(self):
        """"""
        self.detachBtn = cmds.button( l="Detach", c=lambda *args: self.detachCmd() )

    #----------------------------------------------------------------------
    def installExportImport(self):
        """"""
        cmds.frameLayout( l='Export & Import:', bs='etchedIn', cll=False, mh=5, w=200)
        cmds.button(l="Export",c=lambda *args: self.exportWeight() )
        cmds.button(l="Import",c=lambda *args: self.importWeight() )        
        cmds.setParent( ".." )
        
    #----------------------------------------------------------------------
    def editCmd(self):
        """"""
        self.influence = sel = self.__selectInfluence()
        
        if cmds.objectType( self.influence, isType='joint' ):
            self._type =  'joint'
            deformers = cmds.listConnections('%s.wm' % self.influence, d=1, s=0, t='skinCluster', et=1)
            
        elif cmds.listRelatives(self.influence, s=1, typ='clusterHandle'):
            self._type =  'cluster'
            deformers = cmds.listConnections('%s.wm' % self.influence, d=1, s=0, t='cluster', et=1)
        else: cmds.error("%s is unsupported type!" % self.influence)
        
        try:
            infObjs = cmds.listConnections('%s.og' % deformers[0])
            cmds.select(infObjs)
        except TypeError:
            cmds.error("Can't find geometry deformed by %s" % self.influence)
        
        # Get into component select mode
        cmds.selectMode( object=True )
        cmds.selectMode( component=True )
        self.toggleButton()
    
    #----------------------------------------------------------------------
    def doneCmd(self):
        """"""
        if cmds.ls(sl=1):
            softSelectData = scData.SoftSelectionData(None, setup.getSupportTypes().keys())
            weightData = softSelectData.getWeightData()
            append = cmds.checkBox(self.appendCheck, q=1, v=1)
            
            if self._type == 'joint':
                scData.JointFn.redefineWeight(self.influence, weightData, append)
            elif self._type == 'cluster':
                scData.ClusterFn.redefineWeight(self.influence, weightData, append)
        
        # Recover select mode
        cmds.selectMode( object=True )
        cmds.select( self.influence )
        self.toggleButton()
    
    #----------------------------------------------------------------------
    def detachCmd(self):
        """"""
        try:
            sel = cmds.ls(sl=1, ap=1)
            inf = sel.pop(0)
            
            if cmds.objectType( inf, isType='joint' ):
                scData.JointFn.detachFromGeometries(inf, sel)
            elif cmds.listRelatives(inf, s=1, typ='clusterHandle'):            
                scData.ClusterFn.detachFromGeometries(inf, sel)
            else: cmds.error("%s is unsupported type!" % inf)
        except IndexError:
            cmds.error("Please cluster (or joint) first, then geometries you want to detach!")
        
        cmds.select(inf)
        om.MGlobal.displayInfo('Detach %s from %s success.' % (inf, str(sel)))

    #----------------------------------------------------------------------
    def exportWeight(self):
        """"""
        sel = self.__selectInfluence()
        
        fileFilter = "Soft Cluster EX Weight Data (*.scw)"
        filePath = cmds.fileDialog2(fm=0, 
                                   ff=fileFilter,
                                   cap='Export Weight Data',
                                   ds=2,
                                   okc='Export')[0]
            
        if cmds.objectType( sel, isType='joint' ):
            scData.JointFn.exportWeight(sel, filePath)
        elif cmds.listRelatives(sel, s=1, typ='clusterHandle'):            
            scData.ClusterFn.exportWeight(sel, filePath)
        else: cmds.error("%s is unsupported type!" % sel)
        
        om.MGlobal.displayInfo('Export to %s success.' % filePath)
    
    #----------------------------------------------------------------------
    def importWeight(self):
        """"""
        sel = self.__selectInfluence()
        
        fileFilter = "Soft Cluster EX Weight Data (*.scw)"
        filePath = cmds.fileDialog2(fm=1, 
                                   ff=fileFilter,
                                   cap='Import Weight Data',
                                   dialogStyle=2,
                                   okc='Import')[0]
        
        if cmds.objectType( sel, isType='joint' ):
            scData.JointFn.importWeight(sel, filePath)
        elif cmds.listRelatives(sel, s=1, typ='clusterHandle'):
            scData.ClusterFn.importWeight(sel, filePath)
        else: cmds.error("%s is unsupported type!" % sel)

        om.MGlobal.displayInfo('Import weight from %s success.' % filePath)
    
    #----------------------------------------------------------------------
    def toggleButton(self):
        """"""
        self.toggle = not self.toggle
        cmds.button(self.editBtn, e=1, en=self.toggle)
        cmds.button(self.doneBtn, e=1, en=not self.toggle)
    
    #----------------------------------------------------------------------
    def __selectInfluence(self):
        """"""
        try:
            return cmds.ls(sl=1, ap=1)[0]
        except IndexError:
            cmds.error("Please select cluster or joint!")


#----------------------------------------------------------------------
def getTab():
    """"""
    return TabWeight()